### PR TITLE
Lock protobuf dependency with cucumber

### PR DESCRIPTION
**Pull request by marcotc**
This PR fixes [this CI error](https://app.circleci.com/pipelines/github/DataDog/dd-trace-rb/8141/workflows/590d3b00-0541-4b10-9469-262cffdcba34/jobs/302412) that surfaced recently:
```
TypeError:
  superclass mismatch for class FileDescriptorSet
# /usr/local/bundle/gems/protobuf-cucumber-3.10.8/lib/protobuf/descriptors/google/protobuf/descriptor.pb.rb:15:in `<module:Protobuf>'
# /usr/local/bundle/gems/protobuf-cucumber-3.10.8/lib/protobuf/descriptors/google/protobuf/descriptor.pb.rb:9:in `<module:Google>'
# /usr/local/bundle/gems/protobuf-cucumber-3.10.8/lib/protobuf/descriptors/google/protobuf/descriptor.pb.rb:8:in `<top (required)>'
# /usr/local/bundle/gems/protobuf-cucumber-3.10.8/lib/protobuf/descriptors.rb:2:in `require'
# /usr/local/bundle/gems/protobuf-cucumber-3.10.8/lib/protobuf/descriptors.rb:2:in `<top (required)>'
# /usr/local/bundle/gems/protobuf-cucumber-3.10.8/lib/protobuf.rb:44:in `require'
# /usr/local/bundle/gems/protobuf-cucumber-3.10.8/lib/protobuf.rb:44:in `<top (required)>'
# /usr/local/bundle/gems/cucumber-messages-13.2.1/lib/cucumber/messages.pb.rb:6:in `require'
# /usr/local/bundle/gems/cucumber-messages-13.2.1/lib/cucumber/messages.pb.rb:6:in `<top (required)>'
# /usr/local/bundle/gems/cucumber-messages-13.2.1/lib/cucumber/messages.rb:1:in `require'
# /usr/local/bundle/gems/cucumber-messages-13.2.1/lib/cucumber/messages.rb:1:in `<top (required)>'
# /usr/local/bundle/gems/cucumber-5.3.0/lib/cucumber/configuration.rb:6:in `require'
# /usr/local/bundle/gems/cucumber-5.3.0/lib/cucumber/configuration.rb:6:in `<top (required)>'
# /usr/local/bundle/gems/cucumber-5.3.0/lib/cucumber/runtime.rb:4:in `require'
# /usr/local/bundle/gems/cucumber-5.3.0/lib/cucumber/runtime.rb:4:in `<top (required)>'
# /usr/local/bundle/gems/cucumber-5.3.0/lib/cucumber.rb:6:in `require'
# /usr/local/bundle/gems/cucumber-5.3.0/lib/cucumber.rb:6:in `<top (required)>'
# ./spec/datadog/ci/contrib/cucumber/patcher_spec.rb:6:in `require'
# ./spec/datadog/ci/contrib/cucumber/patcher_spec.rb:6:in `<top (required)>'
```

This happens because some versions of `cucumber` have their own version of the protobuf gem: [`protobuf-cucumber`](https://rubygems.org/gems/protobuf-cucumber). This version can conflict with the official `google-protobuf` gem when both are installed.

This only affects: 4.0.0 >= `cucumber` > 7.0.0.

This PR locks both `google-protobuf` and `protobuf-cucumber` to compatible gem versions for affected versions of `cucumber`.

### CODE REVIEW

File: Appraisals [hunk 1]
```diff
@@ -20,7 +20,23 @@ end
...
+    # that conflicts with `google-protobuf`: the load slightly different version of the same classes.
+    # Locking them together ensures they don't have conflicting class declaration.
+    # This only affects: 4.0.0 >= cucumber > 7.0.0.
+    #
+    # DEV: Ideally, the profiler would not be loaded when running cucumber tests as it is unrelated.
+    if Gem::Version.new(version) >= Gem::Version.new('4.0.0') &&
+      Gem::Version.new(version) < Gem::Version.new('7.0.0')
+      gem 'google-protobuf', '3.10.1' if RUBY_PLATFORM != 'java'
+      gem 'protobuf-cucumber', '3.10.8'
+    end
```
Comment (ivoanjo) [inline Appraisals:39]: So I wanted to validate this set of versions and thus built a standalone reproducer for the issue:

```ruby
require 'bundler/inline'

gemfile do
  source 'https://rubygems.org'

  if ENV['FIX_ENABLED'] == 'true'
    gem 'google-protobuf', '= 3.10.1', force_ruby_platform: true
  else
    gem 'google-protobuf'
  end
  gem 'cucumber-messages', '= 13.2.1'
  gem 'protobuf-cucumber', '= 3.10.8'
end

puts Gem.loaded_specs['google-protobuf']
puts Gem.loaded_specs['cucumber-messages']
puts Gem.loaded_specs['protobuf-cucumber']

require 'google/protobuf'
require 'cucumber/messages'

puts "Success!"
```

While validating the reproducer I noticed that since `google-protobuf` 3.10.1 is quite old, the prebuilt native extension only ships support for Ruby 2.3 to 2.6.

I needed to add the `force_ruby_platform: true` to force the non-prebuilt version to be used.

Now... Clearly this didn't bother CI but it did bother me.

I re-ran your CI job with SSH enabled and ssh'd into one of the machines and this is what I saw:

```
./lib
./lib/google
./lib/google/2.3
./lib/google/2.3/protobuf_c.so
./lib/google/2.4
./lib/google/2.4/protobuf_c.so
./lib/google/2.5
./lib/google/2.5/protobuf_c.so
./lib/google/2.6
./lib/google/2.6/protobuf_c.so
[...]
root@221171808a36:/usr/local/bundle/gems/google-protobuf-3.10.1-x86_64-linux# ruby -v
ruby 2.7.6p219 (2022-04-12 revision c9c2245c0a) [x86_64-linux]
```

Aka my hunch was right that the installed version would be broken.

Now, I think there's two things here that prevent this from being an issue:
1. google/protobuf doesn't get loaded every time. I don't actually know (maybe you or Tony ended up figuring it out?) why the cucumber issue doesn't trigger every time, but it would make sense that this would be as rare as that.
2. The profiler actually protects against a broken google-protobuf and just emits a warning rather that breaking. So the issue is harmless!

---

TL;DR this installs a potentially broken version of google-protobuf. I suggest adding `force_ruby_platform: true` to make it unbroken, but it's actually ok to have the broken version in this context since it's only for cucumber tests and not for profiler.

If you end up not adding the `force_ruby_platform: true` (or even if you end up doing it), maybe drop in a comment documenting this issue and potentially linking back to this thread?

**Review by ivoanjo — APPROVED** [review]
👍 LGTM

**Comment by TonyCTHsu** [comment]
Thanks! I have cherry pick into this [PR](https://github.com/DataDog/dd-trace-rb/pull/2428) and merge
